[core] Ruleset references that are relative to another ruleset cannot be resolved

<!-- Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core]. -->

**Affects PMD Version:** 

7.26

**Description:**

PMD's RuleSetLoader can't resolve rule file references, if both these conditions are met:

- the referenced file is not on the classpath
- the referenced file's relative path cannot be resolved from the working dir

Consequence: Rule file references that are relative to the rule file from which they're referenced from cannot be resolved.
This mostly happens due to Gradle's usage of PMD (absolute path to main ruleset, running from another workdir), I still think it would be sensible to also resolve sibling rulesets.

This causes the bug described in https://github.com/gradle/gradle/issues/24885.

**Exception Stacktrace:**

From the testcase below:

```
net.sourceforge.pmd.lang.rule.RuleSetLoadException: Cannot load ruleset base.xml: Cannot resolve rule/ruleset reference 'base.xml'.  Make sure the resource is a valid file or URL and is on the CLASSPATH. Use --debug (or a fine log level) to see the current classpath.

	at net.sourceforge.pmd.lang.rule.RuleSetFactory.readDocument(RuleSetFactory.java:203)
	at net.sourceforge.pmd.lang.rule.RuleSetFactory.createRuleSet(RuleSetFactory.java:119)
	at net.sourceforge.pmd.lang.rule.RuleSetFactory.createRuleSet(RuleSetFactory.java:114)
	at net.sourceforge.pmd.lang.rule.RuleSetFactory.parseRuleSetReferenceNode(RuleSetFactory.java:389)
	at net.sourceforge.pmd.lang.rule.RuleSetFactory.parseRuleNode(RuleSetFactory.java:335)
	at net.sourceforge.pmd.lang.rule.RuleSetFactory.parseRulesetNode(RuleSetFactory.java:242)
	at net.sourceforge.pmd.lang.rule.RuleSetFactory.readDocument(RuleSetFactory.java:186)
	at net.sourceforge.pmd.lang.rule.RuleSetFactory.createRuleSet(RuleSetFactory.java:119)
	at net.sourceforge.pmd.lang.rule.RuleSetFactory.createRuleSet(RuleSetFactory.java:114)
	at net.sourceforge.pmd.lang.rule.RuleSetLoader.loadFromResource(RuleSetLoader.java:248)
	at net.sourceforge.pmd.lang.rule.RuleSetLoader.loadFromResource(RuleSetLoader.java:139)
	at net.sourceforge.pmd.lang.rule.RuleSetFactoryTest.testRefToSiblingRuleset(RuleSetFactoryTest.java:62)
Caused by: java.io.FileNotFoundException: Cannot resolve rule/ruleset reference 'base.xml'.  Make sure the resource is a valid file or URL and is on the CLASSPATH. Use --debug (or a fine log level) to see the current classpath.
	at net.sourceforge.pmd.lang.rule.internal.RuleSetReferenceId.notFoundException(RuleSetReferenceId.java:309)
	at net.sourceforge.pmd.lang.rule.internal.RuleSetReferenceId.getInputStream(RuleSetReferenceId.java:304)
	at net.sourceforge.pmd.lang.rule.RuleSetFactory.readDocument(RuleSetFactory.java:165)
```

**Code Sample demonstrating the issue:**

Add the following to net.sourceforge.pmd.lang.rule.RuleSetFactoryTest:

```java
@Test
void testRefToSiblingRuleset(@TempDir Path tempDir) throws IOException {
    write(tempDir.resolve("leaf.xml"), rulesetXml(dummyRule()));
    write(tempDir.resolve("base.xml"), rulesetXml(rulesetRef("leaf.xml")));
    write(tempDir.resolve("main.xml"), rulesetXml(rulesetRef("base.xml")));

    RuleSet rs = new RuleSetLoader().loadFromResource(tempDir.resolve("main.xml").toString());
    assertEquals(1, rs.size());
    assertNotNull(rs.getRuleByName("MockRuleName"));
}

private static void write(Path file, String content) throws IOException {
    Files.write(file, content.getBytes(StandardCharsets.UTF_8));
}
```

**Steps to reproduce:**

Run the (failing) test pasted above.

Alternatively, try the setup linked in the Gradle issue: https://github.com/lukelukes/gradle-pmd-bug

**Running PMD through:** *Gradle*

<!-- If relevant, also include your JDK and OS information, e.g. for ClassNotFoundException, LinkageError, reflection failures, etc. -->

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -58,6 +58,7 @@ Note: Type data is not yet accessible in XPath rules or the PMD Rule Designer. T
     * [#4952](https://github.com/pmd/pmd/issues/4952): \[doc] Improve doc around PMDConfiguration#prependAuxclasspath #setClassloader
     * [#4953](https://github.com/pmd/pmd/issues/4953): \[core] Deprecate PMDConfiguration#setClassloader and #getClassloader
     * [#6865](https://github.com/pmd/pmd/issues/6865): \[core] Include the running PMD version in the "Unable to find referenced rule" error
+    * [#6952](https://github.com/pmd/pmd/issues/6952): \[core] Ruleset references are not resolved relative to the referencing ruleset
 * java
     * [#5041](https://github.com/pmd/pmd/issues/5041): \[java] Parsing failed in ParseLock#doParse(): IndexOutOfBoundsException 
     * [#6768](https://github.com/pmd/pmd/issues/6768): \[java] Disambiguation IllegalStateException resolving a synthesized record accessor used as a call argument alongside an anonymous class

--- a/pmd-ant/src/test/java/net/sourceforge/pmd/ant/PMDTaskTest.java
+++ b/pmd-ant/src/test/java/net/sourceforge/pmd/ant/PMDTaskTest.java
@@ -82,6 +82,17 @@ class PMDTaskTest extends AbstractAntTest {
         }
     }
 
+    /** Gradle uses PMD's Ant integration, this is a test for <a href="https://github.com/gradle/gradle/issues/24885">gradle/gradle#24885</a> */
+    @Test
+    void testRelativeRulesetReference() throws IOException {
+        executeTarget("testRelativeRulesetReference");
+
+        try (InputStream in = Files.newInputStream(Paths.get("target/pmd-ant-relative-ruleset.txt"))) {
+            assertThat(IOUtil.readToString(in, StandardCharsets.UTF_8),
+                       containsString("sample.dummy:1:\tSampleXPathRule:\tTest Rule 2"));
+        }
+    }
+
     @Test
     void testXmlFormatter() throws IOException {
         executeTarget("testXmlFormatter");

--- a/pmd-ant/src/test/resources/net/sourceforge/pmd/ant/rulesets/base.xml
+++ b/pmd-ant/src/test/resources/net/sourceforge/pmd/ant/rulesets/base.xml
@@ -1,0 +1,9 @@
+<?xml version="1.0"?>
+<ruleset name="Base Ruleset" xmlns="http://pmd.sourceforge.net/ruleset/2.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
+
+    <description>Referenced by main.xml with a plain file name.</description>
+
+    <!-- note: this ruleset is in pmd-core/src/test/resources -->
+    <rule ref="rulesets/dummy/basic.xml/SampleXPathRule" />
+</ruleset>

--- a/pmd-ant/src/test/resources/net/sourceforge/pmd/ant/rulesets/main.xml
+++ b/pmd-ant/src/test/resources/net/sourceforge/pmd/ant/rulesets/main.xml
@@ -1,0 +1,8 @@
+<?xml version="1.0"?>
+<ruleset name="Main Ruleset" xmlns="http://pmd.sourceforge.net/ruleset/2.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
+
+    <description>References a sibling ruleset by its plain file name.</description>
+
+    <rule ref="base.xml" />
+</ruleset>

--- a/pmd-ant/src/test/resources/net/sourceforge/pmd/ant/xml/pmdtasktest.xml
+++ b/pmd-ant/src/test/resources/net/sourceforge/pmd/ant/xml/pmdtasktest.xml
@@ -39,6 +39,17 @@
         </pmd>
     </target>
 
+    <!-- https://github.com/pmd/pmd/issues/6952 -->
+    <target name="testRelativeRulesetReference">
+        <pmd noCache="true" failOnError="true">
+            <ruleset>${pmd.home}/src/test/resources/net/sourceforge/pmd/ant/rulesets/main.xml</ruleset>
+            <formatter type="text" toFile="${pmd.home}/target/pmd-ant-relative-ruleset.txt" />
+            <fileset dir="${pmd.home}/src/test/resources/net/sourceforge/pmd/ant/src">
+                <include name="sample.dummy"/>
+            </fileset>
+        </pmd>
+    </target>
+
     <target name="testXmlFormatter">
         <pmd noCache="true">
             <!-- note: this ruleset is in pmd-core/src/test/resources -->

--- a/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleSetFactory.java
+++ b/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleSetFactory.java
@@ -331,6 +331,7 @@ final class RuleSetFactory {
             String ref = REF.getAttributeOrThrow(ruleNode, err);
             RuleSetReferenceId refId = parseReferenceAndWarn(ref, REF.getAttributeNode(ruleNode), err);
             if (refId != null) {
+                refId = RuleSetReferenceId.resolveSibling(ruleSetReferenceId, refId);
                 if (refId.isAllRules()) {
                     parseRuleSetReferenceNode(ruleSetBuilder, ruleNode, ref, refId, rulesetReferences, err);
                 } else {

--- a/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/internal/RuleSetReferenceId.java
+++ b/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/internal/RuleSetReferenceId.java
@@ -4,6 +4,9 @@
 
 package net.sourceforge.pmd.lang.rule.internal;
 
+import static java.util.Objects.requireNonNull;
+
+import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
@@ -187,6 +190,20 @@ public class RuleSetReferenceId {
                 ruleName = null;
             }
         }
+    }
+
+    /**
+     * Resolves {@code ref} against the directory of {@code parent}, if that yields an existing file.
+     * Otherwise, it is returned unchanged, to be resolved as usual (CWD, URL, classpath).
+     */
+    public static RuleSetReferenceId resolveSibling(RuleSetReferenceId parent, RuleSetReferenceId ref) {
+        // isAbsolute() checks ruleSetReference != null and getRuleSetFileName() just returns ruleSetReference, so it can't return null
+        File parentDir = parent.isAbsolute() ? new File(requireNonNull(parent.getRuleSetFileName())).getParentFile() : null;
+        if (parentDir == null || !ref.isAbsolute() || new File(requireNonNull(ref.getRuleSetFileName())).isAbsolute()) {
+            return ref;
+        }
+        File sibling = new File(parentDir, ref.getRuleSetFileName());
+        return sibling.isFile() ? new RuleSetReferenceId(sibling.getPath(), ref.getRuleName()) : ref;
     }
 
     public @Nullable RuleSetReferenceId getParentRulesetIfThisIsARule() {

--- a/pmd-core/src/test/java/net/sourceforge/pmd/lang/rule/RuleSetFactoryTest.java
+++ b/pmd-core/src/test/java/net/sourceforge/pmd/lang/rule/RuleSetFactoryTest.java
@@ -18,11 +18,16 @@ import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
+import java.io.IOException;
 import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.HashSet;
 import java.util.Set;
 
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
 import org.mockito.Mockito;
 
 import net.sourceforge.pmd.properties.PropertyDescriptor;
@@ -42,6 +47,25 @@ class RuleSetFactoryTest extends RulesetFactoryTestBase {
 
         rs = new RuleSetLoader().loadFromResource(TEST_RULESET_1);
         assertEquals(rs.getFileName(), TEST_RULESET_1, "wrong RuleSet file name");
+    }
+
+    /**
+     * @see <a href="https://github.com/pmd/pmd/issues/6952">pmd/pmd#6952</a>
+     */
+    @Test
+    void testRefToSiblingRuleset(@TempDir Path tempDir) throws IOException {
+        write(tempDir.resolve("leaf.xml"), rulesetXml(dummyRule()));
+        write(tempDir.resolve("base.xml"), rulesetXml(rulesetRef("leaf.xml")));
+        write(tempDir.resolve("main.xml"), rulesetXml(rulesetRef("base.xml")));
+
+        // the rulesets are neither on the classpath nor in the working directory
+        RuleSet rs = new RuleSetLoader().loadFromResource(tempDir.resolve("main.xml").toString());
+        assertEquals(1, rs.size());
+        assertNotNull(rs.getRuleByName("MockRuleName"));
+    }
+
+    private static void write(Path file, String content) throws IOException {
+        Files.write(file, content.getBytes(StandardCharsets.UTF_8));
     }
 
     @Test

